File: CollectionLayout-OC/CollectionLayout/ControllerGroup/AlignmentViewController.m
//
//  AlignmentViewController.m
//  PinterestDemo
//
//  Created by 苏沫离 on 2018/9/8.
//  Copyright © 2020 苏沫离. All rights reserved.
//

#define CellIdentifer @"CollectionAlignmentCell"

#import "AlignmentViewController.h"
#import "CollectionAlignmentCell.h"
#import "YLCollectionAlignmentLayout.h"
#import "CollectionSectionHeaderView.h"

@interface AlignmentViewController ()
<UICollectionViewDelegate,UICollectionViewDataSource,YLCollectionAlignmentLayoutDelegate>
@property (nonatomic ,strong) UICollectionView *collectionView;
@property (nonatomic, strong) UIButton *rightButton;
@end

@implementation AlignmentViewController

#pragma mark - life cycyle

- (void)viewDidLoad {
    [super viewDidLoad];
    [self.view addSubview:self.collectionView];
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithCustomView:self.rightButton];
}

- (void)viewWillLayoutSubviews{
    [super viewWillLayoutSubviews];
    self.collectionView.frame = self.view.bounds;
}

#pragma mark - response click

- (void)rightBarButtonItemClick:(UIButton *)sender{
    YLCollectionAlignmentLayout *flowLayout = (YLCollectionAlignmentLayout *)self.collectionView.collectionViewLayout;
    if ([sender.titleLabel.text isEqualToString:@"右对齐"]) {
        [sender setTitle:@"左对齐" forState:UIControlStateNormal];
        flowLayout.cellAlignmentType = YLCollectionAlignmentLeft;
    }else if ([sender.titleLabel.text isEqualToString:@"左对齐"]){
        [sender setTitle:@"右对齐" forState:UIControlStateNormal];
        flowLayout.cellAlignmentType = YLCollectionAlignmentRight;
    }
}

#pragma mark - YLAlignmentFlowLayoutDelegate

-(CGSize)collectionViewElementOfKind:(NSString *)elementOfKind referenceSizeInSection:(NSInteger)section{
    if ([elementOfKind isEqualToString:UICollectionElementKindSectionHeader]) {
        return CGSizeMake(CGRectGetWidth(UIScreen.mainScreen.bounds), 172 / 375.0 * CGRectGetWidth(UIScreen.mainScreen.bounds));
    }
    return CGSizeZero;
}

- (CGSize)collectionViewSizeForItemAtIndexPath:(NSIndexPath *)indexPath{
    return DataModel.shareDemoData[indexPath.row].alignmentTitleSize;
}

#pragma mark - UICollectionViewDelegate

- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView{
    return 3;
}

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section{
    return DataModel.shareDemoData.count;
}

- (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView viewForSupplementaryElementOfKind:(NSString *)kind atIndexPath:(NSIndexPath *)indexPath{
    if ([kind isEqualToString:UICollectionElementKindSectionHeader]) {
        CollectionSectionHeaderView *headerView = [collectionView dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:kCollectionSectionHeaderIdentifer forIndexPath:indexPath];
        return headerView;
    }
    CollectionSectionFooterView *footerView = [collectionView dequeueReusableSupplementaryViewOfKind:UICollectionElementKindSectionFooter withReuseIdentifier:kCollectionSectionFooterIdentifer forIndexPath:indexPath];
    return footerView;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath{
    CollectionAlignmentCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:CellIdentifer forIndexPath:indexPath];
    cell.itemLable.text = DataModel.shareDemoData[indexPath.row].title;
    return cell;
}

#pragma mark - setter and getter

- (UIButton *)rightButton{
    if (_rightButton == nil) {
        UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
        button.titleLabel.font = [UIFont systemFontOfSize:16];
        [button setTitleColor:UIColor.redColor forState:UIControlStateNormal];
        [button setTitle:@"左对齐" forState:UIControlStateNormal];
        [button addTarget:self action:@selector(rightBarButtonItemClick:) forControlEvents:UIControlEventTouchUpInside];
        _rightButton = button;
    }
    return _rightButton;
}

- (UICollectionView *)collectionView{
    if (_collectionView == nil){
        YLCollectionAlignmentLayout *flowLayout = [[YLCollectionAlignmentLayout alloc] init];
         flowLayout.alignmentDelegate = self;
         flowLayout.cellAlignmentType = YLCollectionAlignmentLeft;
         flowLayout.minimumLineSpacing = 16;//行间距
         flowLayout.minimumInteritemSpacing = 16;//列间距
         flowLayout.sectionInset = UIEdgeInsetsMake(0, 12, 20, 12);

        UICollectionView *collectionView = [[UICollectionView alloc] initWithFrame:CGRectMake(0, 0,CGRectGetWidth(self.view.frame), CGRectGetHeight(self.view.frame)) collectionViewLayout:flowLayout];
        collectionView.showsVerticalScrollIndicator = NO;
        collectionView.showsHorizontalScrollIndicator = NO;
        collectionView.delegate = self;
        collectionView.dataSource = self;
        collectionView.backgroundColor = [UIColor whiteColor];
        [collectionView registerClass:[CollectionAlignmentCell class] forCellWithReuseIdentifier:CellIdentifer];
        [collectionView registerClass:CollectionSectionHeaderView.class forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:kCollectionSectionHeaderIdentifer];
        [collectionView registerClass:CollectionSectionFooterView.class forSupplementaryViewOfKind:UICollectionElementKindSectionFooter withReuseIdentifier:kCollectionSectionFooterIdentifer];
        _collectionView = collectionView;
    }

    return _collectionView;
}

@end

File: CollectionLayout-Swift/CollectionLayout-Swift/ControllerGroup/AlignmentViewController.swift
//
//  AlignmentViewController.swift
//  CollectionLayout-Swift
//
//  Created by 苏沫离 on 2020/9/11.
//  Copyright © 2020 苏沫离. All rights reserved.
//
// 左对齐、右对齐

import UIKit

class AlignmentViewController: UIViewController,UICollectionViewDelegate,
                               UICollectionViewDataSource,
                               YLCollectionAlignmentLayoutDelegate {
    ///懒加载
    lazy var rightButton : UIButton = {
        let button = UIButton(type: .custom)
        button.titleLabel?.font = UIFont.systemFont(ofSize: 16)
        button.setTitleColor(UIColor.red, for: .normal)
        button.setTitle("左对齐", for: .normal)
        button.addTarget(self, action: #selector(rightBarButtonItemClick(sender:)), for: .touchUpInside)
        return button
    }()
    
    lazy var collectionView : UICollectionView! = {
        let flowLayout = YLCollectionAlignmentLayout()
        flowLayout.alignmentDelegate = self
        flowLayout.cellAlignmentType = YLCollectionAlignment.left
        flowLayout.minimumLineSpacing = 16 //行间距
        flowLayout.minimumInteritemSpacing = 16 //列间距
        flowLayout.sectionInset = UIEdgeInsets.init(top: 0, left: 12, bottom: 20, right: 12)

        let collectionView = UICollectionView(frame: self.view.bounds, collectionViewLayout: flowLayout)
        collectionView.showsVerticalScrollIndicator = false
        collectionView.showsHorizontalScrollIndicator = false
        collectionView.delegate = self
        collectionView.dataSource = self
        collectionView.backgroundColor = UIColor.white
        collectionView.register(CollectionAlignmentCell.self, forCellWithReuseIdentifier: "CollectionAlignmentCell")
        collectionView.register(CollectionSectionHeaderView.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "HeaderView")
        collectionView.register(CollectionSectionFooterView.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionFooter, withReuseIdentifier: "footerView")
        return collectionView
    }()

    
    // MARK: - life cycyle
    override func viewDidLoad() {
        super.viewDidLoad()

        self.view.addSubview(collectionView)
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(customView: rightButton)
    }
    
    override func viewWillLayoutSubviews() {
        super.viewWillLayoutSubviews()
        collectionView.frame = self.view.bounds
    }
    
    
    // MARK: - response click
    @objc func rightBarButtonItemClick(sender : UIButton) {
        let flowLayout : YLCollectionAlignmentLayout = collectionView.collectionViewLayout as! YLCollectionAlignmentLayout
        
        if sender.titleLabel?.text == "右对齐" {
            sender.setTitle("左对齐", for: .normal)
            flowLayout.cellAlignmentType = YLCollectionAlignment.left
        }else if sender.titleLabel?.text == "左对齐" {
            sender.setTitle("右对齐", for: .normal)
            flowLayout.cellAlignmentType = YLCollectionAlignment.Right
        }
    }
    
    // MARK: - YLAlignmentFlowLayoutDelegate
    
    func collectionViewElementOfKind(elementOfKind: String, referenceSizeInSection: NSInteger) -> CGSize {
        if elementOfKind == UICollectionView.elementKindSectionHeader {
            return CGSize(width: UIScreen.main.bounds.width, height: 172 / 375.0 * UIScreen.main.bounds.width)
        }else{
            return CGSize.zero
        }
    }
    
    func collectionViewSizeForItemAtIndexPath(indexPath: NSIndexPath) -> CGSize {
        return DataModel.shareDemoData()[indexPath.row].alignmentTitleSize;
    }

    // MARK: - UICollectionViewDelegate
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return 3
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return DataModel.shareDemoData().count
    }
    
    func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView {
        if kind == UICollectionView.elementKindSectionHeader {
            let headerView = collectionView.dequeueReusableSupplementaryView(ofKind: UICollectionView.elementKindSectionHeader, withReuseIdentifier: "HeaderView", for: indexPath)
            return headerView
        }
        let footerView = collectionView.dequeueReusableSupplementaryView(ofKind: UICollectionView.elementKindSectionFooter, withReuseIdentifier: "footerView", for: indexPath)
        return footerView
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell : CollectionAlignmentCell = collectionView.dequeueReusableCell(withReuseIdentifier: "CollectionAlignmentCell", for: indexPath) as! CollectionAlignmentCell
        cell.itemLable.text = DataModel.shareDemoData()[indexPath.row].title as String?
        return cell
    }
    
}


